schema-layer/data-structures/vector.md

```diff
@@ -0,0 +1,361 @@
+1:            [a b c]                    [d  e  f]
+         ┌─────┘ │ └─────┐        ┌───────┘  │  └────────┐
+0:    [1 2 3] [4 5 6] [7 8 9] [10 11 12] [13 14 15] [16 17 18]
+
+      Size: 18
+      Max size before root overflow: 27
+      Head chain: A → a → 1
+      Tail chain: B → f(full) → 18(full)
+```
+
+At 27 items, we have a full `height` `2`, `width` `3` tree:
+
+```
+2:                                       [A  B  C]
+                ┌─────────────────────────┘  │  └─────────────────────────────┐
+1:            [a b c]                    [d  e  f]                        [g  h  i]
+         ┌─────┘ │ └─────┐        ┌───────┘  │  └────────┐         ┌───────┘  │  └────────┐
+0:    [1 2 3] [4 5 6] [7 8 9] [10 11 12] [13 14 15] [16 17 18] [19 20 21] [22 23 24] [25 26 27]
+
+      Size: 27
+      Max size before root overflow: 27
+      Head chain: A → a → 1
+      Tail chain: C(full) → i(full) → 27(full)
+```
+
+An overflow adds an additional level and space for `width`<sup>`height+1`</sup>, or `3`<sup>`4`</sup>`= 81` items.
+
+```
+3:                                                                                              [i  ii]
+                                             ┌───────────────────────────────────────────────────┘  │
+2:                                       [A  B  C]                                                 [D]
+                ┌─────────────────────────┘  │  └─────────────────────────────┐                     │
+1:            [a b c]                    [d  e  f]                        [g  h  i]                [j]
+         ┌─────┘ │ └─────┐        ┌───────┘  │  └────────┐         ┌───────┘  │  └────────┐         │
+0:    [1 2 3] [4 5 6] [7 8 9] [10 11 12] [13 14 15] [16 17 18] [19 20 21] [22 23 24] [25 26 27] [28 29 30]
+
+      Size: 30
+      Max size before root overflow: 81
+      Head chain: i → A → a → 1
+      Tail chain: ii → D → j → 30(full)
+```
+
+### Algorithmic properties
+
+We can derive the **maximum number of elements containable under a node and its children** with a `height` of `X` with `width`<sup>`X + 1`</sup>. At a `height` of `0`, we can only fit `width` elements into a single node. Add a `height` `1` parent and we can fit in `width`<sup>`2`</sup> elements across all child `height` `0` nodes underneath it. This calculation is useful for at least `Size` and append (which we will call `PushTail`) operations.
+
+**Overflows** during `PushTail` operations do not always cascade to upper levels of the tree. Cascades only occur where an upper level is maximally full. So a cascade causing the addition of a new root level at `height + 1` only happens when all levels are full.
+
+**All nodes are valid IPLD Vectors**, and as such may be extracted to represent slices of the overall parent data structure. This becomes less useful as we increase `width` as slices must align with node boundaries, particularly the head of the slice (as the tail may be adjusted with minimal mutations).
```

Comment:
OK, so I've tried to address this as:

> This becomes less useful as we increase `width`, as the start index of a slice must align with a node boundary such that the new head node is a copy of an existing node. Otherwise all elements need to be shuffled toward the head, therefore mutating the entire data structures. Lack of alignment a the tail of a slice only requires mutations on the tail chain.